File: sdk/typescript/src/chaos/llm-faults.ts
/**
 * llm-faults.ts — LLM Provider 故障模拟
 *
 * 对齐 Go 端 internal/chaos/llm_faults.go
 * Stability: Experimental
 */

import type { Fault, CleanupFunc } from './faults.js';

/** LLM HTTP 状态码故障（对齐 Go LLMHTTPStatusFault） */
export class LLMHTTPStatusFault implements Fault {
  readonly provider: string;
  readonly statusCode: number;
  readonly body: string;
  readonly durationMs: number;
  private _active = false;

  constructor(provider: string, statusCode: number, body: string, durationMs = 30000) {
    this.provider = provider;
    this.statusCode = statusCode;
    this.body = body;
    this.durationMs = durationMs;
  }

  type(): string { return `llm_http_${this.statusCode}`; }

  description(): string {
    return `模拟 ${this.provider} Provider 返回 HTTP ${this.statusCode}`;
  }

  async inject(): Promise<CleanupFunc> {
    this._active = true;
    return async () => { this._active = false; };
  }

  get active(): boolean { return this._active; }
}

/** 创建 503 故障 */
export function llmHTTP503Fault(provider: string): LLMHTTPStatusFault {
  return new LLMHTTPStatusFault(
    provider, 503,
    '{"error": {"message": "Service Unavailable", "type": "server_error"}}',
  );
}

/** 创建 429 限流故障 */
export function llmHTTP429Fault(provider: string): LLMHTTPStatusFault {
  return new LLMHTTPStatusFault(
    provider, 429,
    '{"error": {"message": "Rate limit exceeded", "type": "rate_limit_error"}}',
  );
}

/** 创建 500 服务器错误 */
export function llmHTTP500Fault(provider: string): LLMHTTPStatusFault {
  return new LLMHTTPStatusFault(
    provider, 500,
    '{"error": {"message": "Internal Server Error", "type": "server_error"}}',
  );
}

/** LLM 超时故障（对齐 Go LLMTimeoutFault） */
export class LLMTimeoutFault implements Fault {
  readonly provider: string;
  readonly timeoutMs: number;
  private _active = false;

  constructor(provider: string, timeoutMs: number) {
    this.provider = provider;
    this.timeoutMs = timeoutMs;
  }

  type(): string { return 'llm_timeout'; }

  description(): string {
    return `模拟 ${this.provider} Provider 响应超时 ${this.timeoutMs}ms`;
  }

  async inject(): Promise<CleanupFunc> {
    this._active = true;
    return async () => { this._active = false; };
  }

  get active(): boolean { return this._active; }
}

/** LLM 间歇性故障（对齐 Go LLMIntermittentFault） */
export class LLMIntermittentFault implements Fault {
  readonly provider: string;
  readonly failureRate: number;
  readonly failureStatus: number;
  private _active = false;

  constructor(provider: string, failureRate: number, failureStatus = 503) {
    this.provider = provider;
    this.failureRate = failureRate;
    this.failureStatus = failureStatus;
  }

  type(): string { return 'llm_intermittent'; }

  description(): string {
    return `模拟 ${this.provider} Provider 间歇性故障（故障率 ${(this.failureRate * 100).toFixed(0)}%）`;
  }

  async inject(): Promise<CleanupFunc> {
    this._active = true;
    return async () => { this._active = false; };
  }

  get active(): boolean { return this._active; }

  /** 判断本次请求是否应触发故障 */
  shouldFail(): boolean {
    return Math.random() < this.failureRate;
  }
}

/** LLM 慢响应故障（对齐 Go LLMSlowResponseFault） */
export class LLMSlowResponseFault implements Fault {
  readonly provider: string;
  readonly minDelayMs: number;
  readonly maxDelayMs: number;
  private _active = false;

  constructor(provider: string, minDelayMs: number, maxDelayMs: number) {
    this.provider = provider;
    this.minDelayMs = minDelayMs;
    this.maxDelayMs = maxDelayMs;
  }

  type(): string { return 'llm_slow_response'; }

  description(): string {
    return `模拟 ${this.provider} Provider 慢响应 ${this.minDelayMs}ms~${this.maxDelayMs}ms`;
  }

  async inject(): Promise<CleanupFunc> {
    this._active = true;
    return async () => { this._active = false; };
  }

  get active(): boolean { return this._active; }

  /** 计算本次延迟时间 */
  computeDelay(): number {
    if (this.maxDelayMs <= this.minDelayMs) return this.minDelayMs;
    return this.minDelayMs + Math.random() * (this.maxDelayMs - this.minDelayMs);
  }
}

// ===== 预定义故障场景 =====

/** LLM 故障场景（对齐 Go LLMFaultScenario） */
export interface LLMFaultScenario {
  name: string;
  provider: string;
  faults: Fault[];
}

/** 创建 LLM 故障转移场景：503 → 429 → 超时 → 恢复 */
export function llmFailoverScenario(provider: string): LLMFaultScenario {
  return {
    name: 'llm_failover_sequence',
    provider,
    faults: [
      llmHTTP503Fault(provider),
      llmHTTP429Fault(provider),
      new LLMTimeoutFault(provider, 5000),
    ],
  };
}

/** 创建 LLM 混沌场景：间歇故障 + 慢响应 */
export function llmChaosScenario(provider: string): LLMFaultScenario {
  return {
    name: 'llm_chaos_mixed',
    provider,
    faults: [
      new LLMIntermittentFault(provider, 0.3),
      new LLMSlowResponseFault(provider, 1000, 5000),
    ],
  };
}

// ===== 向后兼容别名 =====

/** @deprecated 使用 LLMHTTPStatusFault */
export class LLMErrorFault implements Fault {
  readonly statusCode: number;
  readonly errorMessage: string;
  readonly modelTarget: string;
  private _active = false;

  constructor(statusCode: number, errorMessage: string, modelTarget: string, _probability = 1.0) {
    this.statusCode = statusCode;
    this.errorMessage = errorMessage;
    this.modelTarget = modelTarget;
  }

  type(): string { return 'llm_error'; }
  description(): string { return `LLM ${this.modelTarget} 返回 ${this.statusCode}: ${this.errorMessage}`; }
  async inject(): Promise<CleanupFunc> {
    this._active = true;
    return async () => { this._active = false; };
  }
}

/** @deprecated 使用 LLMTimeoutFault */
export class LLMRateLimitFault implements Fault {
  readonly maxRequests: number;
  readonly windowMs: number;
  readonly modelTarget: string;
  private _active = false;
  private requestCount = 0;
  private windowStart = Date.now();

  constructor(maxRequests: number, windowMs: number, modelTarget: string) {
    this.maxRequests = maxRequests;
    this.windowMs = windowMs;
    this.modelTarget = modelTarget;
  }

  type(): string { return 'llm_rate_limit'; }
  description(): string { return `LLM ${this.modelTarget} 限流 ${this.maxRequests} req/${this.windowMs}ms`; }
  async inject(): Promise<CleanupFunc> {
    this._active = true;
    return async () => { this._active = false; this.requestCount = 0; };
  }

  shouldThrottle(): boolean {
    if (!this._active) return false;
    const now = Date.now();
    if (now - this.windowStart > this.windowMs) {
      this.windowStart = now;
      this.requestCount = 0;
    }
    this.requestCount++;
    return this.requestCount > this.maxRequests;
  }
}

File: agentprimordia/internal/chaos/llm_faults.go
package chaos

import (
	"context"
	"fmt"
	"math/rand"
	"net/http"
	"sync/atomic"
	"time"
)

// ===== LLM Provider 故障模拟 =====

// LLMHTTPStatusFault LLM HTTP 状态码故障
type LLMHTTPStatusFault struct {
	Provider   string // Provider 名称
	StatusCode int    // HTTP 状态码
	Body       string // 响应体
	Duration   time.Duration
	server     *http.Server
	active     atomic.Bool
}

// LLMHTTP503Fault 创建 503 故障
func LLMHTTP503Fault(provider string) *LLMHTTPStatusFault {
	return &LLMHTTPStatusFault{
		Provider:   provider,
		StatusCode: 503,
		Body:       `{"error": {"message": "Service Unavailable", "type": "server_error"}}`,
		Duration:   30 * time.Second,
	}
}

// LLMHTTP429Fault 创建 429 限流故障
func LLMHTTP429Fault(provider string) *LLMHTTPStatusFault {
	return &LLMHTTPStatusFault{
		Provider:   provider,
		StatusCode: 429,
		Body:       `{"error": {"message": "Rate limit exceeded", "type": "rate_limit_error"}}`,
		Duration:   30 * time.Second,
	}
}

// LLMHTTP500Fault 创建 500 server error
func LLMHTTP500Fault(provider string) *LLMHTTPStatusFault {
	return &LLMHTTPStatusFault{
		Provider:   provider,
		StatusCode: 500,
		Body:       `{"error": {"message": "Internal Server Error", "type": "server_error"}}`,
		Duration:   30 * time.Second,
	}
}

func (f *LLMHTTPStatusFault) Type() string {
	return fmt.Sprintf("llm_http_%d", f.StatusCode)
}

func (f *LLMHTTPStatusFault) Description() string {
	return fmt.Sprintf("模拟 %s Provider 返回 HTTP %d", f.Provider, f.StatusCode)
}

func (f *LLMHTTPStatusFault) Inject(ctx context.Context) (CleanupFunc, error) {
	// 启动一个模拟故障的 HTTP 服务器
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/chat/completions", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(f.StatusCode)
		w.Write([]byte(f.Body))
	})
	mux.HandleFunc("/v1/completions", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(f.StatusCode)
		w.Write([]byte(f.Body))
	})

	f.server = &http.Server{
		Addr:    ":18999", // 使用非标准端口避免冲突
		Handler: mux,
	}

	f.active.Store(true)
	go func() {
		_ = f.server.ListenAndServe()
	}()

	return func(ctx context.Context) error {
		if f.active.Load() {
			f.active.Store(false)
			shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()
			return f.server.Shutdown(shutdownCtx)
		}
		return nil
	}, nil
}

// LLMTimeoutFault LLM 超时故障
type LLMTimeoutFault struct {
	Provider  string
	Timeout   time.Duration
	active    atomic.Bool
	server    *http.Server
}

// LLMTimeoutFault 创建超时故障
func NewLLMTimeoutFault(provider string, timeout time.Duration) *LLMTimeoutFault {
	return &LLMTimeoutFault{
		Provider: provider,
		Timeout:  timeout,
	}
}

func (f *LLMTimeoutFault) Type() string { return "llm_timeout" }

func (f *LLMTimeoutFault) Description() string {
	return fmt.Sprintf("模拟 %s Provider 响应超时 %v", f.Provider, f.Timeout)
}

func (f *LLMTimeoutFault) Inject(ctx context.Context) (CleanupFunc, error) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/chat/completions", func(w http.ResponseWriter, r *http.Request) {
		// 模拟超时：等待比request timeout更长的时间
		select {
		case <-time.After(f.Timeout + 5*time.Second):
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{"choices":[{"message":{"content":"late response"}}]}`))
		case <-r.Context().Done():
			// 客户端已超时断开
			return
		}
	})

	f.server = &http.Server{
		Addr:    ":18998",
		Handler: mux,
	}

	f.active.Store(true)
	go func() {
		_ = f.server.ListenAndServe()
	}()

	return func(ctx context.Context) error {
		if f.active.Load() {
			f.active.Store(false)
			shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()
			return f.server.Shutdown(shutdownCtx)
		}
		return nil
	}, nil
}

// LLMIntermittentFault LLM 间歇性故障
type LLMIntermittentFault struct {
	Provider      string
	FailureRate   float64 // 0.0~1.0，故障概率
	FailureStatus int     // 故障时的 HTTP 状态码
	active        atomic.Bool
	server        *http.Server
}

// NewLLMIntermittentFault 创建间歇性故障
func NewLLMIntermittentFault(provider string, failureRate float64) *LLMIntermittentFault {
	return &LLMIntermittentFault{
		Provider:      provider,
		FailureRate:   failureRate,
		FailureStatus: 503,
	}
}

func (f *LLMIntermittentFault) Type() string { return "llm_intermittent" }

func (f *LLMIntermittentFault) Description() string {
	return fmt.Sprintf("模拟 %s Provider 间歇性故障（故障率 %.0f%%）", f.Provider, f.FailureRate*100)
}

func (f *LLMIntermittentFault) Inject(ctx context.Context) (CleanupFunc, error) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/chat/completions", func(w http.ResponseWriter, r *http.Request) {
		if rand.Float64() < f.FailureRate {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(f.FailureStatus)
			w.Write([]byte(`{"error": {"message": "Intermittent failure", "type": "server_error"}}`))
			return
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		w.Write([]byte(`{"choices":[{"message":{"content":"ok"}}]}`))
	})

	f.server = &http.Server{
		Addr:    ":18997",
		Handler: mux,
	}

	f.active.Store(true)
	go func() {
		_ = f.server.ListenAndServe()
	}()

	return func(ctx context.Context) error {
		if f.active.Load() {
			f.active.Store(false)
			shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()
			return f.server.Shutdown(shutdownCtx)
		}
		return nil
	}, nil
}

// LLMSlowResponseFault LLM 慢响应故障
type LLMSlowResponseFault struct {
	Provider  string
	MinDelay  time.Duration
	MaxDelay  time.Duration
	active    atomic.Bool
	server    *http.Server
}

// NewLLMSlowResponseFault 创建慢响应故障
func NewLLMSlowResponseFault(provider string, minDelay, maxDelay time.Duration) *LLMSlowResponseFault {
	return &LLMSlowResponseFault{
		Provider: provider,
		MinDelay: minDelay,
		MaxDelay: maxDelay,
	}
}

func (f *LLMSlowResponseFault) Type() string { return "llm_slow_response" }

func (f *LLMSlowResponseFault) Description() string {
	return fmt.Sprintf("模拟 %s Provider 慢响应 %v~%v", f.Provider, f.MinDelay, f.MaxDelay)
}

func (f *LLMSlowResponseFault) Inject(ctx context.Context) (CleanupFunc, error) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/chat/completions", func(w http.ResponseWriter, r *http.Request) {
		delay := f.MinDelay
		if f.MaxDelay > f.MinDelay {
			delay += time.Duration(rand.Int63n(int64(f.MaxDelay - f.MinDelay)))
		}
		time.Sleep(delay)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		w.Write([]byte(`{"choices":[{"message":{"content":"slow response"}}]}`))
	})

	f.server = &http.Server{
		Addr:    ":18996",
		Handler: mux,
	}

	f.active.Store(true)
	go func() {
		_ = f.server.ListenAndServe()
	}()

	return func(ctx context.Context) error {
		if f.active.Load() {
			f.active.Store(false)
			shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()
			return f.server.Shutdown(shutdownCtx)
		}
		return nil
	}, nil
}

// LLMFaultScenario 预定义 LLM 故障场景序列
type LLMFaultScenario struct {
	Name      string
	Provider  string
	Faults    []Fault
}

// LLMFailoverScenario 创建完整的 LLM 故障转移场景
// 模拟：503 → 429 → 超时 → 恢复
func LLMFailoverScenario(provider string) *LLMFaultScenario {
	return &LLMFaultScenario{
		Name:     "llm_failover_sequence",
		Provider: provider,
		Faults: []Fault{
			LLMHTTP503Fault(provider),
			LLMHTTP429Fault(provider),
			NewLLMTimeoutFault(provider, 5*time.Second),
		},
	}
}

// LLMChaosScenario 创建 LLM 混沌场景（间歇故障 + 慢响应）
func LLMChaosScenario(provider string) *LLMFaultScenario {
	return &LLMFaultScenario{
		Name:     "llm_chaos_mixed",
		Provider: provider,
		Faults: []Fault{
			NewLLMIntermittentFault(provider, 0.3),
			NewLLMSlowResponseFault(provider, 1*time.Second, 5*time.Second),
		},
	}
}
